kadmin/mod.c

```diff
@@ -303,3 +303,270 @@ mod_entry(struct modify_options *opt, int argc, char **argv)
+        if (ret)
+            break;
+        *nkstuple = etypes.len;
+        *kstuple = ecalloc(etypes.len, sizeof(kstuple[0][0]));
+        for (i = 0; i < etypes.len; i++) {
+            (*kstuple)[i].ks_enctype = etypes.val[i];
+            (*kstuple)[i].ks_salttype = 0;
+        }
+        return 0;
+    }
+    if (princ->n_key_data > 0) {
+        *kstuple = ecalloc(1, sizeof(kstuple[0][0]));
+        *nkstuple = 1;
+        for (i = 0; i < princ->n_key_data; i++) {
+            if (princ->key_data->key_data_kvno == kvno) {
+                (*kstuple)[0].ks_enctype = princ->key_data->key_data_type[0];
+                (*kstuple)[0].ks_salttype = princ->key_data->key_data_type[1];
+                return 0;
+            }
+        }
+    }
+    krb5_warnx(context, "Could not determine what enctypes to generate "
+               "keys for; recreate namespace?");
+    return EINVAL;
+}
+
+static int
+randkey_kr(kadm5_principal_ent_rec *princ,
+           unsigned int old_kvno,
+           unsigned int kvno)
+{
+    krb5_key_salt_tuple *kstuple = 0;
+    krb5_error_code ret = 0;
+    size_t nkstuple = 0;
+
+    /*
+     * We might be using kadm5clnt, so we'll use kadm5_randkey_principal_3(),
+     * which will generate new keys on the server side.  This allows a race,
+     * but it will be detected by the key rotation update checks in lib/kadm5
+     * and lib/hdb.
+     */
+    ret = princ2kstuple(princ, old_kvno, &kstuple, &nkstuple);
+    if (ret == 0)
+        ret = kadm5_randkey_principal_3(kadm_handle, princ->principal, 1,
+                                        nkstuple, kstuple, NULL, NULL);
+    free(kstuple);
+    return ret;
+}
+
+static int
+do_mod_ns_kr(krb5_principal principal, void *data)
+{
+    krb5_error_code ret;
+    kadm5_principal_ent_rec princ;
+    struct modify_namespace_key_rotation_options *e = data;
+    HDB_Ext_KeyRotation existing;
+    HDB_Ext_KeyRotation new_kr;
+    HDB_extension ext;
+    KeyRotation new_krs[3];
+    krb5_tl_data *tl;
+    krb5_data d;
+    time_t now = time(NULL);
+    size_t size;
+    int freeit = 0;
+
+    d.data = 0;
+    d.length = 0;
+    new_kr.len = 0;
+    new_kr.val = new_krs;
+    ext.mandatory = 0;
+    ext.data.element = choice_HDB_extension_data_key_rotation;
+    ext.data.u.key_rotation.len = 0;
+    ext.data.u.key_rotation.val = 0;
+    existing.len = 0;
+    existing.val = 0;
+    memset(&new_krs, 0, sizeof(new_krs));
+    memset(&princ, 0, sizeof(princ));
+
+    if (e->force_flag || e->revoke_old_string) {
+        krb5_warnx(context, "--force and --revoke-old not implemented yet");
+        return ENOTSUP;
+    }
+
+    ret = kadm5_get_principal(kadm_handle, principal, &princ,
+			      KADM5_PRINCIPAL | KADM5_KVNO |
+                              KADM5_KEY_DATA | KADM5_TL_DATA);
+    if (ret == 0) {
+        freeit = 1;
+        for (tl = princ.tl_data; tl; tl = tl->tl_data_next) {
+            if (tl->tl_data_type != KRB5_TL_KRB5_CONFIG)
```

Comment:
```suggestion
            if (tl->tl_data_type != KRB5_TL_KEY_ROTATION)
```
or something like that.